helix-core/src/main/java/org/apache/helix/controller/dataproviders/BaseControllerDataProvider.java

```diff
@@ -798,25 +799,18 @@ private void updateDisabledInstances(Collection<InstanceConfig> instanceConfigs,
     _disabledInstanceSet.clear();
     for (InstanceConfig config : instanceConfigs) {
       Map<String, List<String>> disabledPartitionMap = config.getDisabledPartitionsMap();
-      if (!config.getInstanceEnabled()) {
+      if (!InstanceValidationUtil.isInstanceEnabled(config, clusterConfig)) {
         _disabledInstanceSet.add(config.getInstanceName());
       }
       for (String resource : disabledPartitionMap.keySet()) {
-        if (!_disabledInstanceForPartitionMap.containsKey(resource)) {
-          _disabledInstanceForPartitionMap.put(resource, new HashMap<>());
-        }
+        _disabledInstanceForPartitionMap.putIfAbsent(resource, new HashMap<>());
         for (String partition : disabledPartitionMap.get(resource)) {
-          if (!_disabledInstanceForPartitionMap.get(resource).containsKey(partition)) {
-            _disabledInstanceForPartitionMap.get(resource).put(partition, new HashSet<>());
-          }
-          _disabledInstanceForPartitionMap.get(resource).get(partition)
+          _disabledInstanceForPartitionMap.get(resource)
+              .computeIfAbsent(partition, key -> new HashSet<>())
               .add(config.getInstanceName());
         }
       }
     }
-    if (clusterConfig != null && clusterConfig.getDisabledInstances() != null) {
-      _disabledInstanceSet.addAll(clusterConfig.getDisabledInstances().keySet());
-    }
```

Comment:
Is this no longer needed?

Reply:
No longer needed. Everything related to instance disable is integrated into `InstanceValidationUtil.isInstanceEnabled`. It's done in a few lines above.